torch/testing/_internal/mps_utils.py

```diff
@@ -0,0 +1,1012 @@
+COMMON = "TestCommon"
+ERROR_INPUTS = "TestErrorInputs"
+
+
+def xfailUnimplemented(test_func: Callable) -> Callable:
+    def wrapper(*args, **kwargs):  # type: ignore[no-untyped-def]
+        try:
+            test_func(*args, **kwargs)
+        except NotImplementedError as e:
+            raise unittest.SkipTest(
+                "Requires op not currently implemented on MPS"
+            ) from e
+        except TypeError as e:
+            raise unittest.SkipTest("Uses dtype not supported on MPS") from e
+        except unittest.SkipTest as e:
+            # Don't error out on tests that have been explicitly skipped for some other reason
+            raise e
+        except Exception as e:
+            warnings.warn(
+                "Test is marked as unimplemented on MPS,"
+                "but instead of NotImplementedError or TypeError we received {type(e).__name__}:{e} "
+            )
+            raise unittest.SkipTest(
+                "[WARNING]: Test is marked as unimplemented on MPS,"
+                "but instead of NotImplementedError or TypeError we received {type(e).__name__}:{e} "
+            ) from e
+        else:
+            raise RuntimeError(
+                "Test is marked as unimplemented on MPS, but received unexpected success. Ensure CPU fallback is disabled"
+            )
+
+    return wrapper
+
+
+class MPSSkipInfo:
+    def __init__(
+        self,
+        *args: str,
+        test_class: Optional[str] = None,
+        variant: Optional[str] = None,
+        dtypes: Optional[Union[torch.dtype, list[torch.dtype]]] = None,
+        skip: Callable = unittest.expectedFailure,
+        skip_msg: str = "Skipped!",
+        upper: Optional[float] = None,
+        lower: Optional[float] = None,
+    ):
+        """Basic struct for tracking MPS OpInfo xfails
+
+        args: String names of test(s) to apply this xfail info to
+        test_class: Test class, e.g. 'TestCommon' etc.
+        variant: Variant name. Set to empty str ("") to explicitly specify the non-variant case
+        If set to None, will instead apply to all variants of the test
+        dtypes: If none specified, xfails all dtype variants
+        skip: Type of decorator to add [expectedFailure, skipTest, xfailUnimplementedOpMPS, xfailUnimplementedDtypeMPS]
+        upper: Upper bound MacOS version this xfail applies to (exclusive)
+        lower: Lower bound MacOS version this xfail applies to (inclusive)
+        """
+        self.tests: list[str] = []
+        for arg in args:
+            self.tests.append(arg)
+        self.test_class = test_class
+        self.variant = variant
+        self.dtypes = dtypes
+        self.skip = skip
+        self.skip_msg = skip_msg
+        self.upper = upper
+        self.lower = lower
+
+        if UNIMPLEMENTED in self.tests:
+            # Not all tests fail for unimplemented ops, so we specify here
+            self.tests = [TEST_OUTPUT_GRAD_MATCH, TEST_OUTPUT_MATCH]
```

Comment:
Why error inputs are excluded?
```suggestion
            self.tests = [TEST_OUTPUT_GRAD_MATCH, TEST_OUTPUT_MATCH, TEST_ERROR_INPUTS]
```

Reply:
Because I was seeing that the error_inputs test can still return a pass even if the op hasn't been implemented in mps